Introduce the idea of trusted/untrusted snapshot

MVrachev opened the pull request:
Fixes #1523 

**Description of the changes being introduced by the pull request**:

If you do the following steps with a repository that contains hashes in meta dicts:
1. call Updater.refresh() and load, verify and cache all metadata files
2. modify timestamp snapshot meta information:
(One or more of hashes or length for snapshot changes here)
3. call Updater.refresh() again
4. root and timestamp will be updated to their latest versions
5. local snapshot will be loaded, but hashes/length will be different
then the ones in timestamp.snapshot_meta and that will prevent loading
6. remote snapshot is loaded and verification starts

then when executing step 6 the rollback checks will not be done because
the old snapshot was not loaded on step 5.

In order to resolve this issue, we are introducing the idea of trusted and
untrusted snapshot.
Trusted snapshot is the locally available cached version. This version has
been verified at least once meaning hashes and length were already checked
against timestamp.snapshot_meta hashes and length.
That's why we can allow loading a trusted snapshot version even if there is a
mismatch between the current timestamp.snapshot_meta hashes/length and
hashes/length inside the trusted snapshot.
Untrusted snapshot is the one downloaded from the web. It hasn't been verified
before and that's why we mandate that timestamp.snapshot_meta hashes and length
should match the hashes and legth calculated on this untrusted version of
snapshot.

As the TrustedMetadataSet doesn't have information on which snapshot is trusted or
not, so possibly the best solution is to add a new argument "trusted"
to update_snapshot.
Even though this is ugly as the rest of the update functions doesn't
have such an argument, it seems the best solution as it seems to work
in all cases:
- when loading a local snapshot, we know the data has at some point been
trusted (signatures have been checked): it doesn't need to match hashes
now
- if there is no local snapshot and we're updating from remote, the
remote data must match meta hashes in timestamp
- if there is a local snapshot and we're updating from remote, the remote
data must match meta hashes in timestamp

Lastly, I want to point out that  hash checks for metadata files are not
essential to TUF security guarantees: they are just an additional layer of
security that allows us to avoid even parsing json that could be malicious -
we already know the malicious metadata would be stopped at metadata
verification after the parsing.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
Introduce the idea of trusted/untrusted snapshot

If you do the following steps:
1. call Updater.refresh() and load, verify and cache all metadata files
2. modify timestamp snapshot meta information:
(One or more of hashes or length for snapshot changes here)
3. call Updater.refresh() again
4. root and timestamp will be updated to their latest versions
5. local snapshot will be loaded, but hashes/length will be different
than the ones in timestamp.snapshot_meta and that will prevent loading
6. remote snapshot is loaded and verification starts
then when executing step 6 the rollback checks will not be done because
the old snapshot was not loaded on step 5.

In order to resolve this issue, we are introducing the idea of trusted and
untrusted snapshot.
Trusted snapshot is the locally available cached version. This version has
been verified at least once meaning hashes and length were already checked
against timestamp.snapshot_meta hashes and length.
That's why we can allow loading a trusted snapshot version even if there is a
mismatch between the current timestamp.snapshot_meta hashes/length and
hashes/length inside the trusted snapshot.
Untrusted snapshot is the one downloaded from the web. It hasn't been verified
before and that's why we mandate that timestamp.snapshot_meta hashes and length
should match the hashes and legth calculated on this untrusted version of
snapshot.

As the TrustedMetadataSet doesn't have information which snapshot is trusted or
not, so possibly the best solution is to add a new argument "trusted"
to update_snapshot.
Even though this is ugly as the rest of the update functions doesn't
have such an argument, it seems the best solution as it seems to work
in all cases:
- when loading a local snapshot, we know the data has at some point been
trusted (signatures have been checked): it doesn't need to match hashes
now
- if there is no local snapshot and we're updating from remote, the
remote data must match meta hashes in timestamp
- if there is a local snapshot and we're updating from remote, the remote
data must match meta hashes in timestamp

Lastly, I want to point out that  hash checks for metadata files are not
essential to TUF security guarantees: they are just an additional layer of
security that allows us to avoid even parsing json that could be malicious -
we already know the malicious metadata would be stopped at metadata
verification after the parsing.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

jku commented on tests/test_updater_with_simulator.py:
```diff
@@ -152,6 +152,33 @@ def test_keys_and_signatures(self):
+
+        # The new targets should have a lower version than the local trusted one.
+        self.sim.targets.version = 1
+        self.sim.update_snapshot()
+
+        # Calculate the hash of the new modified version of snapshot.
+        digest_object = sslib_hash.digest("sha256")
+        digest_object.update(self.sim._fetch_metadata("snapshot"))
+        new_snapshot_hash = digest_object.hexdigest()
+        # Assign the new hash to meta in timestamp, so that loading the local
+        # snapshot will fail, but loading the next snapshot version will succeed.
+        self.sim.timestamp.snapshot_meta.hashes = {"sha256": new_snapshot_hash}
```
looking at this with fresh eyes, this seems out of place here: our tests should not be this complex. I don't know where the meta file hashing itself should eventually happen but the test should not have to do any of this... The repo (in this case RepositorySimulator) should instead be able to create `meta` dicts that contains length and hashes.

I think the test should just have something like  `self.sim.metafile_hashes = True` in the beginning of the test and then `self.sim.update_snapshot()` should just work without any manual modification. 

This means `RepositorySimulator.update_snapshot()` (and `update_timestamp()`) would include hashes/length in the meta dicts if the `metafile_hashes` attribute is set.

Can you try if this approach is sensible?

jku commented on tuf/ngclient/_internal/trusted_metadata_set.py:
```diff
@@ -271,6 +273,8 @@ def update_snapshot(self, data: bytes) -> None:
 
         Args:
             data: unverified new snapshot metadata as bytes
+            trusted: whether to consider the given data as trusted. Default is
+                False.
```
please try rephrasing this: current text tells the reader nothing on top of the variable name. Try answering the questions "what does trusted mean?" and/or "when should I set it to True?" -- might not be easy given how complex the use case is

jku commented on tests/test_updater_with_simulator.py:
```diff
@@ -152,6 +152,33 @@ def test_keys_and_signatures(self):
         self._run_refresh()
 
+    def test_snapshot_rollback_check_with_missing_local_snapshot(self):
+        # Test triggering snapshot rollback check on a newly downloaded snapshot
+        # when the local snapshot is not loaded due to hash mismatch.
```
the local snapshot does get loaded as intermediate snapshot (as in, we know it won't be valid as final snapshot but we want to load it for the rollback checks)

jku reviewed (changes requested):
I think the commit/PR title is misleading (ans so are some of the comments): the snapshot rollback checks _require_ local snapshot, and do not work without local snapshot. What you are doing is allowing the loading of local snapshot _as intermediate snapshot_ even if the hashes do not match (because, like meta version, the meta hashes reference the new remote snapshot). We want to load the local snapshot even though we already know it's not valid as final because the local snapshot allows us to do rollback checks on the new remote snapshot.

What might be worth mentioning is that the hash checks for metadata files are _not_ essential to TUF security guarantees: they are just an additional layer of security that allows us to avoid even parsing json that could be malicious (we already know the malicious metadata would be stopped at metadata verification after the parsing).

It's a complex situation and because of the complexity we absolutely must be able to clearly express what happens. If we are not able to do that then we should avoid implementing the metafile hashing altogether...

Commit:
Repository_simulator: add a flag to compute hashes

Add an option to calculate the hashes and length for timestamp/snapshot
meta.
This will help to cover more use cases with the repository simulator.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
@jku  I tried to address all of your comments.
First and foremost I changed the commit, pr and test description. 
I know they are lengthy but didn't found another way to describe the situation.
Can you give me some ideas on how to shorten it? 

Second, I experimented with adding hashes and length calculations in the Repository simulator.
What do you think?

Finally, I updated the `trusted` description as requested, but again it's a little long...

jku commented on tests/repository_simulator.py:
```diff
@@ -243,16 +249,64 @@ def _fetch_metadata(self, role: str, version: Optional[int] = None) -> bytes:
 
+    def _compute_hashes_and_length(
+        self, data: bytes, hash_algorithms: Optional[List[str]]
+    ):
```
for a test implementation, this feels complex: I don't think we need exception handling or configurable algorithms (I think you can just hard code the sslib default)

You're probably thinking about possibly implementing this in Metadata API, but I think we should keep this implementation as simple as possible.

jku commented on tests/test_updater_with_simulator.py:
```diff
@@ -152,6 +152,38 @@ def test_keys_and_signatures(self):
+        # the local snapshot, but to have hash match between timestamp.meta and
+        # the next snapshot version.
+        self.compute_metafile_hashes_length = True
+        self.sim.update_timestamp()
```
if you set the option earlier you don't need to call `update_timestamp()` because `update_snapshot()` has already done that (possibly we should rename update_snapshot if it is confusing).

MVrachev replied:
I forgot `update_snapshot()` calls `update_timestamp()` and because I had been manually changing timestamp before and I had to call `update_timestamp()` manually I forgot I don't have to do it now.

MVrachev replied:
It will be good if we have a more descriptive name, but not sure what it could be...

jku reviewed (commented):
Are you sure this actually runs the hash computing code? Can you double check with `--dump` or debug prints or something?

the docstrings are indeed hard, I'll maybe try to suggest something when I'm more caffeinated... otherwise left two code suggestions.

jku commented on tests/repository_simulator.py:
```diff
@@ -243,16 +249,64 @@ def _fetch_metadata(self, role: str, version: Optional[int] = None) -> bytes:
         self.timestamp.version += 1
 
     def update_snapshot(self):
         self.snapshot.meta["targets.json"].version = self.targets.version
```
meta["targets.json"] hashes and length are never set.

Modifying RepositorySimulator is fine if it makes things easier: maybe `self.delegates()` should actually be `self.all_targets()` (that returns targets+delegatedtargets)

MVrachev replied:
If we are using it only in `update_snapshot()` (as it's the case now) I think that makes sense.
If we ever need to access all delegates without `targets` we would change that.

Commit:
Repo simulator: make delegates() to all_targets()

Modify RepositorySimulator function delegates() to all_targets(), so
that all targets can be traversed and updated with one cycle when
calling update_snapshot() (which is the only use case for now for
delegates()).

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
> Are you sure this actually runs the hash computing code? Can you double check with `--dump` or debug prints or something?

Ohh... I didn't run it because I was assigning `self.compute_metafile_hashes_length` to True
instead of `self.sim.compute_metafile_hashes_length` 🤦. 
Great catch!

I think I addressed your comments.

> the docstrings are indeed hard, I'll maybe try to suggest something when I'm more caffeinated... otherwise left two code suggestions.

Will wait for your other review.

jku commented on tests/repository_simulator.py:
```diff
@@ -121,8 +124,10 @@ def snapshot(self) -> Snapshot:
-        for role, md in self.md_delegates.items():
+    def all_targets(self) -> Iterator[Tuple[str, Targets]]:
+        all_targets = self.md_delegates
+        all_targets["targets"] = self.md_targets
```
this modifies self.md_delegates, that's not right. 

I'm not against changing how the metadata is stored in RepositorySimulator -- maybe merging md_targets into md_delegates does make sense... but not like this.

I would just yield "targets", self.md_targets here

jku commented on tests/repository_simulator.py:
```diff
@@ -243,16 +248,42 @@ def _fetch_metadata(self, role: str, version: Optional[int] = None) -> bytes:
+        for algorithm in hash_algorithms:
+            digest_object = sslib_hash.digest(algorithm)
+            digest_object.update(data)
+            hashes[algorithm] = digest_object.hexdigest()
```
loop and array (and even pre-initializing the dict) seem unnecessary

jku commented on tests/repository_simulator.py:
```diff
@@ -243,16 +248,42 @@ def _fetch_metadata(self, role: str, version: Optional[int] = None) -> bytes:
+        if self.compute_metafile_hashes_length:
+            hashes, length = self._compute_hashes_and_length(
+                self._fetch_metadata("snapshot")
+            )
```
this is just a nit but it might look better and would fit into one line if the function argument was `rolename:str` and the function did the fetch.

jku reviewed (changes requested):
I would maybe move the `compute_metafile_hashes_length = True` to the start of the test: we should try to make tests that don't modify the repository in the middle of the test in unrealistic ways and here it seems we should be testing against a repository that includes hashes for all meta versions, not just a single snapshot update.

Can you make sure the client testing plan (document or issue) includes this new repository setting as a configuration we should keep in mind when improving tests (so that this doesn't remain the single test using this configuration)

I left some specific comments, the important ones are the new bug in RepositorySimulator and how the hash computation is still more complex than it needs to be

jku commented:
As a review of my original RepositorySimulator design:

This use case shows a (sort of) disadvantage of signing on demand: since the meta hashes require signatures to be included in the hash, the snapshot  update now requires signing every targets metadata (which in practice means that files will typically be signed twice per refresh(): once to create snapshot, once to serve the actual file). I think signing on demand is still probably a reasonable/correct approach: manually signing sounds like a lot of work we don't want to do in the tests

MVrachev commented:
> I would maybe move the `compute_metafile_hashes_length = True` to the start of the test: we should try to make tests that don't modify the repository in the middle of the test in unrealistic ways and here it seems we should be testing against a repository that includes hashes for all meta versions, not just a single snapshot update.

Agreed. Changed that.

> Can you make sure the client testing plan (document or issue) includes this new repository setting as a configuration we should keep in mind when improving tests (so that this doesn't remain the single test using this configuration)

Yes, left a comment here: https://github.com/theupdateframework/python-tuf/issues/1579#issuecomment-940801497
 
> I left some specific comments, the important ones are the new bug in RepositorySimulator and how the hash computation is still more complex than it needs to be

I think I addressed all of your comments.
IMO the only questionable item left are the comments. 
Will appreciate a review from you as one who writes documentation in a more succinct way.

jku commented on tuf/ngclient/_internal/trusted_metadata_set.py:
```diff
@@ -271,6 +273,10 @@ def update_snapshot(self, data: bytes) -> None:
+            trusted: whether to consider the given data as trusted. The data is
+                trusted if it has been verified by this function. When trusted,
+                the data will be loaded even if there is a hash/length mismatch.
+                Default is False.
```
```suggestion
            trusted: whether data has at some point been verified by
                TrustedMetadataSet as a valid snapshot. Purpose of trusted is
                to allow loading of locally stored snapshot as intermediate
                snapshot even if hashes in current timestamp meta no longer
                match data. Default is False.
```

:shrug: My version is even longer but it does try to explain the purpose of the argument. Up to you.

jku reviewed (approved):
I think this looks correct. 

The case really is kind of a corner case of a corner case so very hard to explain clearly. @joshuagl does this look reasonable to you?

MVrachev commented:
>🤷 My version is even longer but it does try to explain the purpose of the argument. Up to you.

Liked your version a little more.
Added it.
Additionally, I added a return annotation to `_compute_hashes_and_length`.

joshuagl commented on tests/repository_simulator.py:
```diff
@@ -100,6 +100,9 @@ def __init__(self):
         # target downloads are served from this dict
         self.target_files: Dict[str, RepositoryTarget] = {}
 
+        # Whether to compute hashes and legth for meta in snapshot/timestamp
```
```suggestion
        # Whether to compute hashes and length for meta in snapshot/timestamp
```

joshuagl commented on tests/repository_simulator.py:
```diff
@@ -243,16 +247,34 @@ def _fetch_metadata(self, role: str, version: Optional[int] = None) -> bytes:
+        self, role: str
+    ) -> Tuple[Dict[str, str], str]:
+        data = self._fetch_metadata(role)
+        digest_object = sslib_hash.digest(sslib_hash.DEFAULT_HASH_ALGORITHM)
```
Nit: Most likely outside of the scope of this PR, but would it be prudent to have our own default value for the hash algorithm, rather than relying on sslib? (Both here and in metadata.py) Would we notice if this was changed in sslib? What would the impact of such an unnoticed change be?

MVrachev replied:
I will prefer it if we don't discuss this here as will slow the merging of this pr.
I created an issue for that https://github.com/theupdateframework/python-tuf/issues/1613.

joshuagl commented on tuf/ngclient/updater.py:
```diff
@@ -360,7 +360,7 @@ def _load_snapshot(self) -> None:
 
             data = self._download_metadata("snapshot", length, version)
-            self._trusted_set.update_snapshot(data)
+            self._trusted_set.update_snapshot(data, trusted=False)
```
```suggestion
            self._trusted_set.update_snapshot(data)
```

Nit: we defined a default, should we just use it? Is being explicit here a style recommendation?

MVrachev replied:
Even though we have a default value I prefer if we are explicit about this.
That way it's easier to distinguish when we consider the data as `trusted` and when we do not.
In this case, I see the default value more as a defense mechanism that `trusted` will be set.

joshuagl replied:
Not requiring any change here. But if we want to force ourselves to be explicit about how we call this internal API, why have the default?

MVrachev replied:
What do you think @jku? Should we have a default value for `trusted`?

jku replied:
I don't care very much, but defining a default in internal API and then explicitly avoiding using that default as defense mechanism seems weird. So I'd say use the default or don't define it in the first place... (but it's really not a required change)

MVrachev replied:
I checked and if I decide to change the code to mandate `trusted` it will have too many other unrelated changes, so the tests could pass. It doesn't worth it.
I the argument from this function call as Joshua suggested.

joshuagl reviewed (approved):
This looks reasonable to me, thank you both.

jku commented on tests/repository_simulator.py:
```diff
@@ -243,16 +247,34 @@ def _fetch_metadata(self, role: str, version: Optional[int] = None) -> bytes:
 
+    def _compute_hashes_and_length(
+        self, role: str
+    ) -> Tuple[Dict[str, str], str]:
```
return annotation seems wrong for length

MVrachev commented:
The changes I made: 
- followed Joshua's advice and didn't pass `trusted` at all when calling it with downloaded data. 
- fixed the return type annotation error pointed out by Jussi.

MVrachev commented:
I am puzzled why the windows checks failed.
Do we want to restart the run or merge it as it is?

jku commented:
> Update - Currently investigating inability to process Windows jobs. Linux and Mac jobs are not impacted.
> Oct 13, 09:44 UTC

let's try again a little later
